File: backend/app/api/v1/monitor.py
"""系统监控路由（Phase 8.4，api接口清单 §八）：检索性能 / API 用量 / 索引规模 / 资源。

纯读路径，走 AsyncSession；逻辑在 ``services/monitor_service``（无新表/迁移/依赖，
全聚合既有数据源）。窗口参数（仅 retrieval-perf / api-usage）：today / 7d / all。
外部组件（Milvus/ES/Redis/MinIO）失败 → 该字段 None / ``up:false``，端点不 500。
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db
from app.schemas.monitor import (
    ApiUsageResponse,
    FeedbackReportResponse,
    IndexStatsResponse,
    ResourcesResponse,
    RetrievalPerfResponse,
    TraceDetail,
    TraceListResponse,
)
from app.services import monitor_service
from app.services.feedback_service import build_feedback_report

router = APIRouter(prefix="/monitor", tags=["monitor"])


@router.get("/retrieval-perf", response_model=RetrievalPerfResponse)
async def retrieval_perf(
    window: str = Query("today", pattern="today|7d|all"),
    session: AsyncSession = Depends(get_db),
) -> RetrievalPerfResponse:
    """检索性能漏斗：延迟分布 p50/p95 + 召回/精排漏斗均值 + 精排率 + 反馈计数。"""
    return await monitor_service.get_retrieval_perf(session, window)


@router.get("/api-usage", response_model=ApiUsageResponse)
async def api_usage(
    window: str = Query("today", pattern="today|7d|all"),
    session: AsyncSession = Depends(get_db),
) -> ApiUsageResponse:
    """外部 API 用量（查询侧 PG 派生代理，详见响应 ``note``）。"""
    return await monitor_service.get_api_usage(session, window)


@router.get("/index-stats", response_model=IndexStatsResponse)
async def index_stats(
    session: AsyncSession = Depends(get_db),
) -> IndexStatsResponse:
    """各索引/表规模：PG 行数 + Milvus 向量数 + ES 文档数。"""
    return await monitor_service.get_index_stats(session)


@router.get("/resources", response_model=ResourcesResponse)
async def resources(
    session: AsyncSession = Depends(get_db),
) -> ResourcesResponse:
    """基础设施连通与占用（PG/Redis/Milvus/ES/MinIO）。"""
    return await monitor_service.get_resources(session)


# ---- GET /monitor/traces（M41 全链路追溯）----


@router.get("/traces", response_model=TraceListResponse)
async def traces(
    window: str = Query("today", pattern="today|7d|all"),
    limit: int = Query(50, ge=1, le=200),
    session: AsyncSession = Depends(get_db),
) -> TraceListResponse:
    """M41 全链路追溯：请求列表（span 汇总，最新在前）。"""
    return await monitor_service.get_traces(session, window, limit)


@router.get("/traces/{log_id}", response_model=TraceDetail)
async def trace_detail(
    log_id: int,
    session: AsyncSession = Depends(get_db),
) -> TraceDetail:
    """单请求 span 树（新 dict 原样 / 旧行伪 span 合成）。缺失 404。"""
    d = await monitor_service.get_trace(session, log_id)
    if d is None:
        raise HTTPException(status_code=404, detail="检索日志不存在")
    return d


# ---- GET /monitor/feedback-report（M43 反馈闭环）----


@router.get("/feedback-report", response_model=FeedbackReportResponse)
async def feedback_report(
    days: int = Query(default=30, ge=1, le=365),
    session: AsyncSession = Depends(get_db),
) -> FeedbackReportResponse:
    """M43 反馈闭环报告：分类分布 / repo 分组 / 关键词 / 幻觉告警（×M34 enforcement 交叉）。"""
    return FeedbackReportResponse(**await build_feedback_report(session, days=days))

File: frontend-v2/src/api/monitor.ts
/** 系统监控 REST 客户端（对齐后端 Phase 8.4 / schemas/monitor.py）。 */
import { api } from "./client";

export type MonitorWindow = "today" | "7d" | "all";

// ---- GET /monitor/retrieval-perf ----

export interface LatencyMs {
  avg_total: number | null;
  p50_total: number | null;
  p95_total: number | null;
  avg_recall: number | null;
  avg_rerank: number | null;
}

export interface RetrievalFunnel {
  avg_pool: number | null; // RRF 融合池均值
  avg_final: number | null; // 精排后候选均值
}

export interface FeedbackCounts {
  helpful: number;
  not_helpful: number;
}

export interface RetrievalPerf {
  window: string;
  queries: number;
  latency_ms: LatencyMs;
  funnel: RetrievalFunnel;
  rerank_rate: number | null; // 启用精排占比（0~1）
  feedback: FeedbackCounts;
}

// ---- GET /monitor/resources ----

export interface ComponentInfo {
  up: boolean | null;
  detail?: string | null;
  db_size_bytes?: number | null; // postgres
  used_memory_bytes?: number | null; // redis
  keys?: number | null; // redis
  collections?: number | null; // milvus
  rows?: number | null; // milvus
  doc_count?: number | null; // elasticsearch
  size_bytes?: number | null; // elasticsearch store
  asset_bytes?: number | null; // minio
}

export interface Resources {
  status: "healthy" | "degraded";
  components: Record<string, ComponentInfo>;
}

// ---- GET /monitor/api-usage ----

export interface ApiUsage {
  window: string;
  llm_calls: number;
  embedding_query_calls: number;
  rerank_calls: number;
  generated_tokens_est: number;
  indexed_tokens: number;
  note: string;
}

// ---- GET /monitor/index-stats ----

export interface MilvusCollectionStat {
  name: string;
  dim: number | null;
  rows: number | null;
}

export interface PostgresIndexStats {
  code_chunks: number;
  code_chunks_active: number;
  code_chunks_synced_pct: number | null;
  doc_chunks: number;
  doc_chunks_active: number;
  doc_chunks_synced_pct: number | null;
  chunk_relations: number;
  chunk_relations_stale: number;
  call_graph: number;
  call_graph_active: number;
  code_files: number;
  doc_files: number;
  doc_resources: number;
  retrieval_logs: number;
  conversations: number;
  chat_messages: number;
}

export interface MilvusIndexStats {
  strategy: string;
  collections: MilvusCollectionStat[];
}

export interface EsIndexStats {
  index: string;
  doc_count: number | null;
  by_kind: Record<string, number | null>;
}

export interface IndexStats {
  postgres: PostgresIndexStats;
  milvus: MilvusIndexStats;
  elasticsearch: EsIndexStats;
}

// ---- 调用 ----

export const getRetrievalPerf = (win: MonitorWindow = "today") =>
  api.get<RetrievalPerf>("/v1/monitor/retrieval-perf", { params: { window: win } }).then((r) => r.data);

export const getApiUsage = (win: MonitorWindow = "today") =>
  api.get<ApiUsage>("/v1/monitor/api-usage", { params: { window: win } }).then((r) => r.data);

export const getIndexStats = () =>
  api.get<IndexStats>("/v1/monitor/index-stats").then((r) => r.data);

export const getResources = () =>
  api.get<Resources>("/v1/monitor/resources").then((r) => r.data);

// ---- GET /monitor/traces（M41 全链路追溯）----

export interface TraceTokens {
  prompt: number;
  completion: number;
  n_llm_calls: number;
  estimated: boolean;
}

export interface TraceListItem {
  log_id: number;
  query: string;
  mode: string | null;
  agent: string | null;
  total_ms: number | null;
  tokens: TraceTokens | null;
  has_trace: boolean;
  created_at: string | null;
}

export interface TraceList {
  window: string;
  total: number;
  items: TraceListItem[];
}

export interface TraceSpan {
  span_id: number;
  parent_id: number | null;
  kind: string;
  name: string;
  start_ms: number;
  duration_ms: number | null;
  status: string;
  error: string | null;
  tokens: { prompt: number; completion: number; estimated: boolean } | null;
  attrs: Record<string, unknown>;
}

export interface TraceDetail {
  log_id: number;
  query: string;
  mode: string | null;
  legacy: boolean;
  spans: TraceSpan[];
  summary: { total_ms: number; tokens?: TraceTokens; n_spans: number } | null;
  created_at: string | null;
}

export const listTraces = (win: MonitorWindow = "today", limit = 50) =>
  api
    .get<TraceList>("/v1/monitor/traces", { params: { window: win, limit } })
    .then((r) => r.data);

export const getTrace = (logId: number) =>
  api.get<TraceDetail>(`/v1/monitor/traces/${logId}`).then((r) => r.data);
